libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -2374,14 +2387,62 @@ bool AP_Mission::jump_to_landing_sequence(const Location &current_loc)
         }
 
         GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Landing sequence start");
-        _flags.in_landing_sequence = true;
         return true;
     }
 
     GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "Unable to start landing sequence");
     return false;
 }
 
+/*
+   find the closest point on the mission after a DO_RETURN_PATH_START and before DO_LAND_START or landing
+ */
+bool AP_Mission::jump_to_closest_mission_leg(const Location &current_loc)
+{
+    if (_flags.state == MISSION_RUNNING) {
+        // if mission is already running don't switch away from a active landing or return path
+        if (_flags.in_landing_sequence) {
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Landing sequence active");
+            return true;
+
+        } else if (_flags.in_return_path) {
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Return path active");
+            return true;
+        }
+    }
+
+    uint16_t landing_start_index = 0;
+    float min_distance = -1;
+
+    // Go through mission and check each DO_RETURN_PATH_START
+    for (uint16_t i = 1; i < num_commands(); i++) {
+        Mission_Command tmp;
+        if (read_cmd_from_storage(i, tmp) && (tmp.id == MAV_CMD_DO_RETURN_PATH_START)) {
```

Comment:
I worry about potentially reading all commands from storage although I see we already have existing examples of this.  I suppose there is not better alternative.  indexing them into a table somewhere after the mission has been uploaded might not improve anything unless it was done very carefully and would add complexity.

Reply:
Yeah, this is the only way currently.